zug/util.hpp

```diff
@@ -89,29 +90,87 @@ constexpr decltype(auto) invoke_composition(Fns&& fns, Args&&... args)
 }
 
 template <typename Fn, typename... Fns>
-struct composed
+struct composed : std::tuple<Fn, Fns...>
 {
-    std::tuple<Fn, Fns...> fns;
+    using base_t = std::tuple<Fn, Fns...>;
+
+    template <
+        typename TupleFns,
+        std::enable_if_t<
+            !std::is_same<composed, std::decay_t<TupleFns>>::value>* = nullptr>
+    composed(TupleFns&& fns)
+        : base_t{std::forward<TupleFns>(fns)}
+    {}
 
     template <typename... T>
     constexpr decltype(auto) operator()(T&&... xs) &
     {
-        return invoke_composition(fns, std::forward<T>(xs)...);
+        return invoke_composition(as_tuple(), std::forward<T>(xs)...);
     }
 
     template <typename... T>
     constexpr decltype(auto) operator()(T&&... xs) const&
     {
-        return invoke_composition(fns, std::forward<T>(xs)...);
+        return invoke_composition(as_tuple(), std::forward<T>(xs)...);
     }
 
     template <typename... T>
     constexpr decltype(auto) operator()(T&&... xs) &&
     {
-        return invoke_composition(std::move(fns), std::forward<T>(xs)...);
+        return invoke_composition(std::move(*this).as_tuple(),
+                                  std::forward<T>(xs)...);
+    }
+
+    template <typename... T>
+    base_t& as_tuple() &
```

Comment:
I think you do not need the `template <typename... T>` part, probably a copy-paste typo :) Same on the ones below. I am thinking though... should we get rid of this and just passed the `composed` itself to the various operations? I think that everywhere where we are calling `x.as_tuple()` it would also work with just `x`.

Reply:
Yes, this was 50% c+p, 50% sunday morning lazily not checking my code! But as mentioned in another chat - we need the exact tuple type for both `invoke_composition` (`std::tuple_size`) and `comp` (`std::tuple_cat`).